executor/distsql.go

```diff
@@ -450,62 +485,76 @@ func (e *IndexLookUpExecutor) getRetTpsByHandle() []*types.FieldType {
-		SetStreaming(e.indexStreaming).
-		SetFromSessionVars(e.ctx.GetSessionVars()).
-		SetMemTracker(tracker).
-		SetFromInfoSchema(infoschema.GetInfoSchema(e.ctx)).
-		Build()
-	if err != nil {
-		return err
+
+	kvRanges := [][]kv.KeyRange{e.kvRanges}
+	if e.partitionTableMode {
+		kvRanges = e.partitionKVRanges
 	}
 	tps := e.getRetTpsByHandle()
 	idxID := e.getIndexPlanRootID()
-	// Since the first read only need handle information. So its returned col is only 1.
-	result, err := distsql.SelectWithRuntimeStats(ctx, e.ctx, kvReq, tps, e.feedback, getPhysicalPlanIDs(e.idxPlans), idxID)
-	if err != nil {
-		return err
-	}
-	worker := &indexWorker{
-		idxLookup:       e,
-		workCh:          workCh,
-		finished:        e.finished,
-		resultCh:        e.resultCh,
-		keepOrder:       e.keepOrder,
-		batchSize:       initBatchSize,
-		checkIndexValue: e.checkIndexValue,
-		maxBatchSize:    e.ctx.GetSessionVars().IndexLookupSize,
-		maxChunkSize:    e.maxChunkSize,
-		PushedLimit:     e.PushedLimit,
-	}
-	if worker.batchSize > worker.maxBatchSize {
-		worker.batchSize = worker.maxBatchSize
+	results := make([]distsql.SelectResult, 0, 1)
+	for _, kvRange := range kvRanges {
+		var builder distsql.RequestBuilder
+		kvReq, err := builder.SetKeyRanges(kvRange).
+			SetDAGRequest(e.dagPB).
+			SetStartTS(e.startTS).
+			SetDesc(e.desc).
+			SetKeepOrder(e.keepOrder).
+			SetStreaming(e.indexStreaming).
+			SetFromSessionVars(e.ctx.GetSessionVars()).
+			SetMemTracker(tracker).
+			SetFromInfoSchema(infoschema.GetInfoSchema(e.ctx)).
+			Build()
+		if err != nil {
+			return err
+		}
+		// Since the first read only need handle information. So its returned col is only 1.
+		result, err := distsql.SelectWithRuntimeStats(ctx, e.ctx, kvReq, tps, e.feedback, getPhysicalPlanIDs(e.idxPlans), idxID)
+		if err != nil {
+			return err
+		}
+		results = append(results, result)
 	}
+
 	e.idxWorkerWg.Add(1)
 	go func() {
 		defer trace.StartRegion(ctx, "IndexLookUpIndexWorker").End()
-		ctx1, cancel := context.WithCancel(ctx)
-		_, err := worker.fetchHandles(ctx1, result)
-		if err != nil {
-			e.feedback.Invalidate()
-		}
-		cancel()
-		if err := result.Close(); err != nil {
-			logutil.Logger(ctx).Error("close Select result failed", zap.Error(err))
+		for i, result := range results {
+			worker := &indexWorker{
+				idxLookup:       e,
+				workCh:          workCh,
+				finished:        e.finished,
+				resultCh:        e.resultCh,
+				keepOrder:       e.keepOrder,
+				batchSize:       initBatchSize,
+				checkIndexValue: e.checkIndexValue,
+				maxBatchSize:    e.ctx.GetSessionVars().IndexLookupSize,
+				maxChunkSize:    e.maxChunkSize,
+				PushedLimit:     e.PushedLimit,
+			}
+			if worker.batchSize > worker.maxBatchSize {
+				worker.batchSize = worker.maxBatchSize
+			}
+			if e.partitionTableMode {
+				worker.partitionTable = e.prunedPartitions[i]
+			}
+			ctx1, cancel := context.WithCancel(ctx)
+			if _, err := worker.fetchHandles(ctx1, result); err != nil {
```

Comment:
maybe we should test limit n + partition  situation for this~

e.g.  

we build 10 kvReqs, but `limit 1` condition can be reached when first kvReq handled be tableWorkers side and  upper executor `Close` indexLookupExecutor `e.finished` should be closed..

maybe we should Close remain 10-1 results and return quickly

Reply:
You are right. I'll add some test cases for it.

Reply:
I wanted to add a piece of code to check if `worker.finish` is closed here to make this goroutine can exit more quickly, but if we do that, we also have to add more lines of code to care about resource management like close all unclosed `Result` before exit, which may make this logic complex and hard to maintain.
So after second thought, I suggest that we keep the current status here.

Reply:
I'll add a failpoint test for this case.

Reply:
I add a piece of code at the beginning of the loop to check if this executor is closed. 
PTAL @lysu